pkg/go/graph/weighted_graph.go

```diff
@@ -0,0 +1,409 @@
+package graph
+
+import (
+	"fmt"
+	"math"
+	"strings"
+)
+
+const Infinite = math.MaxInt32
+
+var errModelCycle = fmt.Errorf("model cycle")
+var errTupleCycle = fmt.Errorf("operands AND or BUT NOT cannot be involved in a cycle")
+
+type WeightedAuthorizationModelGraph struct {
+	nodes map[string]*WeightedAuthorizationModelNode
+	edges map[string][]*WeightedAuthorizationModelEdge
+}
+
+// NewWeightedAuthorizationModelGraph creates a new WeightedAuthorizationModelGraph.
+func NewWeightedAuthorizationModelGraph() *WeightedAuthorizationModelGraph {
+	return &WeightedAuthorizationModelGraph{
+		nodes: make(map[string]*WeightedAuthorizationModelNode),
+		edges: make(map[string][]*WeightedAuthorizationModelEdge),
+	}
+}
+
+// AddNode adds a node to the graph with optional operationType and weight.
+func (wg *WeightedAuthorizationModelGraph) AddNode(uniqueLabel, label string, nodeType NodeType) {
+	wg.nodes[uniqueLabel] = &WeightedAuthorizationModelNode{uniqueLabel: uniqueLabel, label: label, nodeType: nodeType}
+}
+
+func (wg *WeightedAuthorizationModelGraph) AddEdge(fromID, toID string, edgeType EdgeType, condition string) {
+	fromNode := wg.nodes[fromID]
+	toNode := wg.nodes[toID]
+	edge := &WeightedAuthorizationModelEdge{from: fromNode, to: toNode, edgeType: edgeType, conditionedOn: condition}
+	wg.edges[fromID] = append(wg.edges[fromID], edge)
+}
+
+// AssignWeights assigns weights to all the edges and nodes of the graph.
+func (wg *WeightedAuthorizationModelGraph) AssignWeights() error {
+	visited := make(map[string]bool)
+	ancestorPath := make([]*WeightedAuthorizationModelEdge, 0)
+	tupleCycleDependencies := make(map[string][]*WeightedAuthorizationModelEdge)
+
+	for node := range wg.nodes {
+		if visited[node] {
+			continue
+		}
+
+		tupleCyles, err := wg.calculateNodeWeight(node, visited, ancestorPath, tupleCycleDependencies)
+		if err != nil {
+			return err
+		}
+		if len(tupleCyles) > 0 {
+			return fmt.Errorf("%d tuple cycles found without resolution", len(tupleCyles))
```

Comment:
here you can extract a constant error and wrap it to provide dynamic details, such as cycle count:
```go
var CycleError = errors.New("cycle error")

// ... code removed for brevity

return fmt.Errorf("%w: count %d", CycleError, len(tupleCycles))
```
then dependent packages can check for a `CycleError` by type and get descriptive output:
```go
if (errors.Is(err, CycleError)) {
    fmt.Println(err)
}

// output = "cycle error: count 5"
```

Reply:
noted, making the changes